File: Cindy/App/AppModel.swift
import Foundation
import Observation

/// App-wide state: calibration profile, workout history and the signal config.
@MainActor
@Observable
final class AppModel {
    private(set) var calibration: CalibrationProfile?
    private(set) var history: [WorkoutRecord] = []
    /// Editable workout plan (exercises, targets, duration); persisted as JSON in UserDefaults.
    var plan: WorkoutPlan {
        didSet { AppModel.savePlan(plan) }
    }
    /// Write every saved workout to Apple Health.
    private(set) var syncToHealth: Bool {
        didSet { UserDefaults.standard.set(syncToHealth, forKey: AppModel.syncToHealthKey) }
    }
    /// Nudge towards the next session with a local notification.
    private(set) var remindersEnabled: Bool {
        didSet { UserDefaults.standard.set(remindersEnabled, forKey: AppModel.remindersEnabledKey) }
    }
    /// Debug: write a per-frame CSV for every workout (Documents/DebugLogs).
    var recordWorkouts: Bool {
        didSet { UserDefaults.standard.set(recordWorkouts, forKey: AppModel.recordWorkoutsKey) }
    }
    /// Whether the first-run intro has been shown. Settings can replay it,
    /// so this is only ever set, never cleared.
    private(set) var hasSeenIntro: Bool {
        didSet { UserDefaults.standard.set(hasSeenIntro, forKey: AppModel.hasSeenIntroKey) }
    }
    /// Light/dark override; `.system` follows the device setting.
    var theme: AppTheme {
        didSet { UserDefaults.standard.set(theme.rawValue, forKey: AppModel.themeKey) }
    }
    /// UI and speech language; `.system` follows the device setting.
    var language: AppLanguage {
        didSet {
            UserDefaults.standard.set(language.rawValue, forKey: AppModel.languageKey)
            Localization.apply(language)
        }
    }
    let config = SignalConfig.default
    /// Estimated readiness for the next session; `nil` until it is computed
    /// and when nothing at all is known yet.
    private(set) var readiness: Readiness?
    /// When the next session is suggested; also what the reminder is set to.
    private(set) var nextSession: NextSession?
    let calibrationStore: CalibrationStore
    let historyStore: HistoryStore
    let health: HealthExporter
    let healthMetrics: HealthMetricsReader
    let reminder: WorkoutReminder

    init(calibrationStore: CalibrationStore = CalibrationStore(), historyStore: HistoryStore = HistoryStore(),
         health: HealthExporter = .shared, healthMetrics: HealthMetricsReader = HealthMetricsReader(),
         reminder: WorkoutReminder = .shared) {
        self.calibrationStore = calibrationStore
        self.historyStore = historyStore
        self.health = health
        self.healthMetrics = healthMetrics
        self.reminder = reminder
        self.plan = AppModel.loadPlan()
        self.syncToHealth = UserDefaults.standard.bool(forKey: AppModel.syncToHealthKey)
        self.remindersEnabled = UserDefaults.standard.bool(forKey: AppModel.remindersEnabledKey)
        self.recordWorkouts = UserDefaults.standard.bool(forKey: AppModel.recordWorkoutsKey)
        self.hasSeenIntro = UserDefaults.standard.bool(forKey: AppModel.hasSeenIntroKey)
        self.theme = AppModel.loadSetting(AppModel.themeKey) ?? .system
        self.language = AppModel.loadSetting(AppModel.languageKey) ?? .system
        Localization.apply(language)
        reload()
    }

    var isCalibrated: Bool { calibration?.isComplete(for: plan) == true }

    /// Most recent completed workout (for the result comparison).
    var lastCompletedRecord: WorkoutRecord? { history.first { $0.completed } }

    private static let planKey = "workoutPlan"
    private static let recordWorkoutsKey = "recordWorkouts"
    private static let hasSeenIntroKey = "hasSeenIntro"
    private static let syncToHealthKey = "syncToHealth"
    private static let remindersEnabledKey = "remindersEnabled"
    private static let themeKey = "appTheme"
    private static let languageKey = "appLanguage"

    /// `nonisolated`: it only reads `UserDefaults`, and as a main-actor member
    /// the generic initializer it hands to `flatMap` would have to cross isolation.
    private nonisolated static func loadSetting<Value: RawRepresentable & Sendable>(_ key: String) -> Value? where Value.RawValue == String {
        UserDefaults.standard.string(forKey: key).flatMap(Value.init(rawValue:))
    }

    private static func loadPlan() -> WorkoutPlan {
        guard let data = UserDefaults.standard.data(forKey: planKey),
              let plan = try? JSONDecoder().decode(WorkoutPlan.self, from: data), plan.isValid else { return .cindy }
        return plan
    }

    private static func savePlan(_ plan: WorkoutPlan) {
        if let data = try? JSONEncoder().encode(plan) {
            UserDefaults.standard.set(data, forKey: planKey)
        }
    }

    func markIntroSeen() {
        hasSeenIntro = true
    }

    func reload() {
        calibration = calibrationStore.load()
        history = historyStore.load()
    }

    func save(_ record: WorkoutRecord) {
        try? historyStore.append(record)
        history = historyStore.load()
        Task { await refreshReadiness() }
        guard syncToHealth else { return }
        // Fire and forget: a failed Health write must not cost the user the result.
        Task { try? await health.export(record) }
    }

    /// Recomputes the readiness estimate. Health is only read while the sync
    /// is on — with the switch off Cindy does not touch Health at all and the
    /// estimate rests on the workout history.
    func refreshReadiness() async {
        let metrics = syncToHealth ? await healthMetrics.read() : .none
        readiness = ReadinessEstimator().estimate(history: history, metrics: metrics)
        await refreshReminder()
    }

    // MARK: - Reminder

    /// Recomputes the suggested date and rewrites the pending notification.
    /// A scheduled notification cannot re-evaluate itself, so this runs on
    /// every readiness refresh.
    func refreshReminder() async {
        nextSession = NextSessionPlanner().plan(history: history, readiness: readiness, now: Date())
        guard remindersEnabled, let nextSession else {
            reminder.cancel()
            return
        }
        await reminder.schedule(nextSession)
    }

    /// Turns reminders on, asking for permission first.
    /// Returns false when notifications were refused.
    @discardableResult
    func enableReminders() async -> Bool {
        guard await reminder.requestAuthorization() else { return false }
        remindersEnabled = true
        await refreshReminder()
        return true
    }

    func disableReminders() {
        remindersEnabled = false
        reminder.cancel()
    }

    // MARK: - Apple Health

    var isHealthAvailable: Bool { health.isAvailable }

    /// Workouts still missing from Health (only meaningful while the sync is on).
    var workoutsPendingInHealth: Int { health.pendingCount(in: history) }

    /// Turns the Health sync on, asking for permission first.
    /// Returns false when writing was refused.
    @discardableResult
    func enableHealthSync() async -> Bool {
        try? await health.requestAuthorization()
        guard health.isAuthorized else { return false }
        syncToHealth = true
        await refreshReadiness()
        return true
    }

    func disableHealthSync() {
        syncToHealth = false
        Task { await refreshReadiness() }
    }

    /// Writes the workouts saved before the sync was switched on. Returns how many were added.
    @discardableResult
    func exportHistoryToHealth() async throws -> Int {
        try await health.exportMissing(from: history)
    }

    func delete(_ record: WorkoutRecord) {
        try? historyStore.delete(id: record.id)
        history = historyStore.load()
    }
}

File: android/app/src/main/kotlin/me/raddatz/cindy/app/AppModel.kt
package me.raddatz.cindy.app

import android.app.Application
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.os.LocaleListCompat
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import me.raddatz.cindy.AppContainer
import me.raddatz.cindy.CindyApplication
import me.raddatz.cindy.core.AppLanguage
import me.raddatz.cindy.core.AppTheme
import me.raddatz.cindy.core.WorkoutPlan
import me.raddatz.cindy.core.calibration.CalibrationProfile
import me.raddatz.cindy.core.persistence.WorkoutRecord
import me.raddatz.cindy.health.HealthAvailability
import me.raddatz.cindy.reminder.NotificationPermission
import me.raddatz.cindy.reminder.ReadinessSnapshot
import java.time.Instant

/** The switches of the settings screen, read back from `SettingsStore`. */
data class SettingsState(
    val syncToHealth: Boolean,
    val remindersEnabled: Boolean,
    val recordWorkouts: Boolean,
    val hasSeenIntro: Boolean,
    val theme: AppTheme,
    val language: AppLanguage,
)

/**
 * App-wide state over [AppContainer] (iOS: `AppModel`): the plan, the calibration profile, the
 * history, readiness and the settings with their side effects.
 *
 * Permission dialogs need an activity, so the screens launch them and hand the outcome in (see
 * [enableHealthSync] and [enableReminders]).
 */
class AppModel(val container: AppContainer, private val application: Application) : ViewModel() {
    private val settingsStore = container.settings

    private val _plan = MutableStateFlow(settingsStore.workoutPlan)

    /** Editable workout plan; persisted as JSON under the iOS key. */
    val plan: StateFlow<WorkoutPlan> = _plan.asStateFlow()

    private val _calibration = MutableStateFlow(container.calibrationStore.load())
    val calibration: StateFlow<CalibrationProfile?> = _calibration.asStateFlow()

    /** Newest first. */
    val history: StateFlow<List<WorkoutRecord>> = container.history.records

    val readiness: StateFlow<ReadinessSnapshot> = container.readiness.snapshot

    private val _settings = MutableStateFlow(readSettings())
    val settings: StateFlow<SettingsState> = _settings.asStateFlow()

    /** A complete calibration for the current plan. */
    val isCalibrated: StateFlow<Boolean> = combine(_plan, _calibration) { plan, profile -> profile?.isComplete(plan) == true }
        .stateIn(viewModelScope, SharingStarted.Eagerly, _calibration.value?.isComplete(_plan.value) == true)

    init {
        viewModelScope.launch {
            settingsStore.changes().collect { _settings.value = readSettings() }
        }
    }

    val lastCompletedRecord: WorkoutRecord? get() = container.history.lastCompletedRecord

    private fun readSettings() = SettingsState(
        syncToHealth = settingsStore.syncToHealth,
        remindersEnabled = settingsStore.remindersEnabled,
        recordWorkouts = settingsStore.recordWorkouts,
        hasSeenIntro = settingsStore.hasSeenIntro,
        theme = settingsStore.appTheme,
        language = settingsStore.appLanguage,
    )

    private fun updateSettings() {
        _settings.value = readSettings()
    }

    // Plan, calibration, history

    fun setPlan(plan: WorkoutPlan) {
        if (!plan.isValid) return
        settingsStore.workoutPlan = plan
        _plan.value = plan
    }

    fun markIntroSeen() {
        settingsStore.hasSeenIntro = true
        updateSettings()
    }

    fun reload() {
        _calibration.value = container.calibrationStore.load()
        container.history.reload()
    }

    fun save(record: WorkoutRecord) = container.history.save(record)

    fun delete(record: WorkoutRecord) = container.history.delete(record)

    /** Recomputes readiness and moves the reminder; health data is read only while the sync is on. */
    fun refreshReadiness() {
        viewModelScope.launch { container.readiness.refresh() }
    }

    // Display

    fun setRecordWorkouts(enabled: Boolean) {
        settingsStore.recordWorkouts = enabled
        updateSettings()
    }

    fun setTheme(theme: AppTheme) {
        settingsStore.appTheme = theme
        updateSettings()
        AppCompatDelegate.setDefaultNightMode(theme.nightMode)
    }

    /** Recreates the activity in the new language (AppCompat stores the choice below Android 13). */
    fun setLanguage(language: AppLanguage) {
        settingsStore.appLanguage = language
        updateSettings()
        AppCompatDelegate.setApplicationLocales(language.localeList)
    }

    // Health Connect

    fun healthAvailability(): HealthAvailability = container.health.availability()

    suspend fun healthPermissionsToRequest(): Set<String> = container.health.requestedPermissions()

    /**
     * Turns the sync on once the permission sheet is closed. Returns false when writing workouts
     * was refused (iOS `enableHealthSync`).
     */
    suspend fun enableHealthSync(): Boolean {
        if (!container.health.canWriteWorkouts()) return false
        settingsStore.syncToHealth = true
        updateSettings()
        container.readiness.refresh()
        return true
    }

    fun disableHealthSync() {
        settingsStore.syncToHealth = false
        updateSettings()
        refreshReadiness()
    }

    /** Workouts still missing from Health Connect (only meaningful while the sync is on). */
    fun workoutsPendingInHealth(): Int = container.health.pendingCount(history.value)

    /** Writes the workouts saved before the sync was switched on. @throws me.raddatz.cindy.health.HealthException */
    suspend fun exportHistoryToHealth(): Int = container.health.exportMissing(history.value)

    // Reminder

    /**
     * Turns reminders on once the notification permission was asked for. Returns false when
     * notifications cannot be posted (iOS `enableReminders`).
     */
    suspend fun enableReminders(): Boolean {
        if (!NotificationPermission.canPost(application)) return false
        settingsStore.remindersEnabled = true
        updateSettings()
        container.readiness.refreshReminder()
        return true
    }

    fun disableReminders() {
        settingsStore.remindersEnabled = false
        updateSettings()
        container.reminderScheduler.cancel()
    }

    /** When the pending reminder will fire, read back from WorkManager. */
    suspend fun pendingReminderDate(): Instant? = container.reminderScheduler.pendingDate()

    companion object {
        val Factory: ViewModelProvider.Factory = viewModelFactory {
            initializer {
                val application = this[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY] as CindyApplication
                AppModel(application.container, application)
            }
        }
    }
}

val AppTheme.nightMode: Int
    get() = when (this) {
        AppTheme.SYSTEM -> AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
        AppTheme.LIGHT -> AppCompatDelegate.MODE_NIGHT_NO
        AppTheme.DARK -> AppCompatDelegate.MODE_NIGHT_YES
    }

val AppLanguage.localeList: LocaleListCompat
    get() = languageCode?.let { LocaleListCompat.forLanguageTags(it) } ?: LocaleListCompat.getEmptyLocaleList()

/** The language AppCompat currently applies; also picks up a change made in the system settings. */
fun currentAppLanguage(): AppLanguage {
    val locales = AppCompatDelegate.getApplicationLocales()
    if (locales.isEmpty) return AppLanguage.SYSTEM
    return when (locales[0]?.language) {
        AppLanguage.GERMAN.languageCode -> AppLanguage.GERMAN
        AppLanguage.ENGLISH.languageCode -> AppLanguage.ENGLISH
        else -> AppLanguage.SYSTEM
    }
}
